### PR TITLE
Add activity / workflow task version check

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Propagate activity / workflow task version
* Add activity / workflow task version check

**Why?**
Make activity / workflow task check more accurate

**How did you test it?**
Existing tests

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Review by yux0 — CHANGES_REQUESTED** [review]
click the wrong approve

File: common/tasktoken/token.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	}
+}
+
+func NewActivityTaskToken(
+	namespaceID string,
+	workflowID string,
+	runID string,
+	scheduledEventID int64,
+	activityId string,
+	activityType string,
```
Comment (yycptt) [inline common/tasktoken/token.go:60]: why do we need activityType in the token? for metrics?

File: service/frontend/workflow_handler.go [hunk 2]
```diff
@@ -1247,14 +1250,17 @@ func (wh *WorkflowHandler) RecordActivityTaskHeartbeatById(ctx context.Context,
...
+	taskToken := tasktoken.NewActivityTaskToken(
+		namespaceID.String(),
+		workflowID,
+		runID,
+		common.EmptyEventID,
+		activityID,
+		"",
+		1,
+		nil,
+		common.EmptyVersion,
```
Comment (yux0) [inline service/frontend/workflow_handler.go:1262]: any future plan to use a real version?

**Review by yycptt — APPROVED** [review]

File: service/history/workflow_task_handler_callbacks.go [hunk 3]
```diff
@@ -411,9 +414,12 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
...
 
 	currentWorkflowTask := ms.GetWorkflowTaskByID(token.GetScheduledEventId())
-	if !ms.IsWorkflowExecutionRunning() || currentWorkflowTask == nil || currentWorkflowTask.Attempt != token.Attempt ||
+	if !ms.IsWorkflowExecutionRunning() ||
+		currentWorkflowTask == nil ||
 		currentWorkflowTask.StartedEventID == common.EmptyEventID ||
-		(token.StartedEventId != common.EmptyEventID && token.StartedEventId != currentWorkflowTask.StartedEventID) {
+		(token.StartedEventId != common.EmptyEventID && token.StartedEventId != currentWorkflowTask.StartedEventID) ||
+		currentWorkflowTask.Attempt != token.Attempt ||
+		(token.Version != common.EmptyVersion && token.Version != currentWorkflowTask.Version) {
```
Comment (yux0) [inline service/history/workflow_task_handler_callbacks.go:422]: is it possible token.Version > currentWorkflowTask.Version? I am not sure if a workflow task retry would update ms wt version.

File: common/tasktoken/token.go (hunk 1 above)
Comment (wxing1292) [reply common/tasktoken/token.go:60]: metrics, e.g. https://github.com/search?q=repo%3Atemporalio%2Ftemporal%20ActivityTypeTag&type=code

File: service/frontend/workflow_handler.go (hunk 2 above)
Comment (wxing1292) [reply service/frontend/workflow_handler.go:1262]: no, the *ActivitybyID APIs are on the deprecation path, and the contract does not include clock or version

File: service/history/workflow_task_handler_callbacks.go (hunk 3 above)
Comment (wxing1292) [reply service/history/workflow_task_handler_callbacks.go:422]: token version comes from workflow task, so if workflow task is updated, then corresponding token version should be the newest value

**Review by yux0 — APPROVED** [review]

**Pull request merged**